mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala

```diff
@@ -546,65 +446,83 @@ private[ml] object FeedForwardTopology {
  * Model of Feed Forward Neural Network.
  * Implements forward, gradient computation and can return weights in vector format.
  * @param layerModels models of layers
- * @param topology topology of the network
+ * @param layers topology of the network
  */
 private[ml] class FeedForwardModel private(
     val layerModels: Array[LayerModel],
-    val topology: FeedForwardTopology) extends TopologyModel {
+    val layers: Array[Layer]) extends TopologyModel {
+
+  private var outputs: Array[BDM[Double]] = null
+  private var deltas: Array[BDM[Double]] = null
+
   override def forward(data: BDM[Double]): Array[BDM[Double]] = {
-    val outputs = new Array[BDM[Double]](layerModels.length)
-    outputs(0) = layerModels(0).eval(data)
+    // Initialize output arrays for all layers. Special treatment for InPlace
+    val currentBatchSize = data.cols
+    // TODO: allocate outputs as one big array and then create BDMs from it
+    if (outputs == null || outputs(0).cols != currentBatchSize) {
+      outputs = new Array[BDM[Double]](layerModels.length)
+      var inputSize = data.rows
+      for (i <- 0 until layerModels.length) {
+        layerModels(i) match {
+          case inPlace: InPlace => outputs(i) = outputs(i - 1)
```

Comment:
Since I don't see `InPlace` being mixed in anywhere except `LayerModel`, can we just replace the empty `InPlace` trait with a `inPlace : Boolean` flag defined in the `LayerModel` interface and reduce unnecessary abstraction?